be/src/exec/pipeline/exchange/sink_buffer.cpp

```diff
@@ -0,0 +1,215 @@
+
+            PUniqueId finst_id;
+            finst_id.set_hi(instance_id.hi);
+            finst_id.set_lo(instance_id.lo);
+            _instance_id2finst_id[instance_id.lo] = std::move(finst_id);
+        }
+
+        _num_remaining_eos = 0;
+        for (auto& [_, num] : _num_sinkers) {
+            _num_remaining_eos += num;
+        }
+    }
+}
+
+SinkBuffer::~SinkBuffer() {
+    DCHECK(is_finished());
+
+    for (auto& [_, buffer] : _buffers) {
+        while (!buffer.empty()) {
+            auto& request = buffer.front();
+            request.params->release_finst_id();
+            buffer.pop();
+        }
+    }
+}
+
+void SinkBuffer::add_request(const TransmitChunkInfo& request) {
+    DCHECK(_num_remaining_eos > 0);
+    if (_is_finishing) {
+        request.params->release_finst_id();
+        return;
+    }
+    {
+        auto& instance_id = request.fragment_instance_id;
+        std::lock_guard<std::mutex> l(*_mutexes[instance_id.lo]);
+        _buffers[instance_id.lo].push(request);
+        _try_to_send_rpc(instance_id);
+    }
+}
+
+bool SinkBuffer::is_full() const {
+    // std::queue' read is concurrent safe without mutex
+    // Judgement may not that accurate because we do not known in advance which
+    // instance the data to be sent corresponds to
+    size_t max_buffer_size = config::pipeline_sink_buffer_size * _buffers.size();
+    size_t buffer_size = 0;
+    for (auto& [_, buffer] : _buffers) {
+        buffer_size += buffer.size();
+    }
+    return buffer_size > max_buffer_size;
+}
+
+bool SinkBuffer::is_finished() const {
+    if (!_is_finishing) {
+        return false;
+    }
+
+    return _num_sending_rpc == 0 && _total_in_flight_rpc == 0;
+}
+
+// When all the ExchangeSinkOperator shared this SinkBuffer are cancelled,
+// the rest chunk request and EOS request needn't be sent anymore.
+void SinkBuffer::cancel_one_sinker() {
+    if (--_num_uncancelled_sinkers == 0) {
+        _is_finishing = true;
+    }
+}
+
+void SinkBuffer::_process_send_window(const TUniqueId& instance_id, const int64_t sequence) {
+    auto& sequences = _unprocessed_sequences[instance_id.lo];
+    sequences.insert(sequence);
+    auto& max_processed_sequence = _max_processed_sequences[instance_id.lo];
+    std::unordered_set<int64_t>::iterator it;
+    while ((it = sequences.find(max_processed_sequence + 1)) != sequences.end()) {
+        sequences.erase(it);
+        ++max_processed_sequence;
+    }
+}
+
+void SinkBuffer::_try_to_send_rpc(const TUniqueId& instance_id) {
+    DeferOp decrease_defer([this]() { --_num_sending_rpc; });
+    ++_num_sending_rpc;
+
+    for (;;) {
+        if (_is_finishing) {
+            return;
+        }
+
+        auto& buffer = _buffers[instance_id.lo];
+        int64_t unprocessed_window_size = _request_sequences[instance_id.lo] - _max_processed_sequences[instance_id.lo];
+        if (buffer.empty() || unprocessed_window_size >= config::pipeline_sink_brpc_dop) {
```

Comment:
`unprocessed_window_size >= config::pipeline_sink_brpc_dop`, add some comment above this if stmt

Reply:
ok